source/steamlib/steamlib_main.cpp

```diff
@@ -16,71 +16,202 @@ You should have received a copy of the GNU Lesser General Public
+#include "../qcommon/qcommon.h"
 #include "steamlib_local.h"
+#include <steam/steam_api.h>
 
-namespace WSWSTEAM {
+namespace WSWSTEAM
+{
+typedef void ( *auth_callback_t )( void *, size_t );
+
+struct CSteamCallbacks {
+	auth_callback_t m_callback;
+	uint8 m_AuthTicket[1024];
+	uint32 m_AuthTicketLength;
+	HAuthTicket m_AuthTicketResult;
+
+	CSteamCallbacks() : m_callback( NULL ), m_AuthTicketLength( 0 ), m_AuthTicketResult( k_HAuthTicketInvalid ) {}
+
+	int GetAuthSessionTicket( void ( *callback )( void *, size_t ) )
+	{
+		m_callback = callback;
+		m_AuthTicketResult = SteamUser()->GetAuthSessionTicket( m_AuthTicket, sizeof( m_AuthTicket ), &m_AuthTicketLength );
+		return m_AuthTicketResult != k_HAuthTicketInvalid ? 1 : 0;
+	}
+
+	STEAM_CALLBACK( CSteamCallbacks, OnGetAuthSessionTicketResponse, GetAuthSessionTicketResponse_t );
+	STEAM_CALLBACK( CSteamCallbacks, OnWorkshopItemInstalled, ItemInstalled_t );
+
+	STEAM_CALLBACK( CSteamCallbacks, OnGameServerChangeRequested, GameServerChangeRequested_t);
+};
+static CSteamCallbacks *g_pCallbacks = nullptr;
+
+void CSteamCallbacks::OnGetAuthSessionTicketResponse( GetAuthSessionTicketResponse_t *pParam )
+{
+	if( pParam->m_eResult != k_EResultOK )
+		m_AuthTicketLength = 0;
+
+	if (m_callback)
+	{
+		m_callback(m_AuthTicket, m_AuthTicketLength);
+	}
+}
+
+void CSteamCallbacks::OnWorkshopItemInstalled( ItemInstalled_t *pParam )
+{
+	if( pParam->m_unAppID == SteamUtils()->GetAppID() ) {
+		Com_Printf( "New workshop item installed, please restart!\n" );
+	}
+}
+
+void CSteamCallbacks::OnGameServerChangeRequested( GameServerChangeRequested_t *pParam )
+{
+	char connectCommand[1024];
+	if( pParam->m_rgchPassword[0] ) {
+		char password[65] = { 0 };
+		Info_CleanValue( pParam->m_rgchPassword, password, sizeof( password ) );
+		Q_snprintfz( connectCommand, sizeof( connectCommand ), "connect \"%s@%s\"\n", password, pParam->m_rgchServer );
+	} else {
+		Q_snprintfz( connectCommand, sizeof( connectCommand ), "connect \"%s\"\n", pParam->m_rgchServer );
+	}
+
+	WSWSTEAM::GetSteamImport()->Cbuf_ExecuteText( EXEC_APPEND, connectCommand );
+}
+
+void AddUGCItem( PublishedFileId_t workshopItemID )
+{
+	uint32 unItemState = SteamUGC()->GetItemState( workshopItemID );
+	if( !( unItemState & k_EItemStateInstalled ) )
+		return;
+
+	uint32 unTimeStamp = 0;
+	uint64 unSizeOnDisk = 0;
+	char szItemFolder[1024] = { 0 };
+
+	if( !SteamUGC()->GetItemInstallInfo( workshopItemID, &unSizeOnDisk, szItemFolder, sizeof( szItemFolder ), &unTimeStamp ) ) {
+		Com_Printf( S_COLOR_RED "Unable to retrieve info for %uLL\n", workshopItemID );
+		return;
+	}
+
+	if( unItemState & k_EItemStateLegacyItem ) {
+		// szItemFolder just points directly to the item for legacy items that were published with the RemoteStorage API.
+		// figure out how to handle this later!
+		Com_Printf( S_COLOR_RED "Skipping legacy item '%s'\n", szItemFolder );
+		return;
+	}
+
+	Com_Printf( "Found workshop folder '%s'\n", szItemFolder );
```

Comment:
We might wanna remove this debug print before shipping?

Reply:
Should there be a cvar incase people want to debug?